### PR TITLE
Implement step 3 of XHR's SetRequestHeader method

### LINKED ISSUE #9548: Implement Step 3 of the SetRequestHeader method in XHR
This is where we should normalize the value being passed in as a `ByteString` paramemter.

Code: `components/script/dom/xmlhttprequest.rs`
Spec: https://xhr.spec.whatwg.org/#dom-xmlhttprequest-setrequestheader, https://fetch.spec.whatwg.org/#concept-header-value-normalize

### PR DESCRIPTION
**Pull request by timvandermeij**
Fixes #9548.

Feedback on this PR is appreciated as I suspect that this can be improved. I have compiled Servo and tested visiting various websites and did not notice any problems. However, a run of the full test suite would be good to have.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/9595)

### CODE REVIEW

**Comment by TimNN** [comment]
I'm not (yet) very familiar with servo but this will panic if `value` is not valid `UTF-8`, which is probably not a good idea (although I don't know if it is possible for a non-UTF-8 `ByteString` to reach this code path).

This will also *always* allocate, even if no trimming needs to be done (not sure if that is a concern).

File: components/script/dom/xmlhttprequest.rs [hunk 1]
```diff
@@ -370,8 +371,11 @@ impl XMLHttpRequestMethods for XMLHttpRequest {
         if self.ready_state.get() != XMLHttpRequestState::Opened || self.send_flag.get() {
             return Err(Error::InvalidState); // Step 1, 2
         }
+        // Step 3
+        let whitespace: &[char] = &['\x09', '\x0A', '\x0D', '\x20'];
+        value = ByteString::from_str(value.as_str().unwrap().trim_matches(whitespace)).unwrap();
         if !name.is_token() || !value.is_field_value() {
-            return Err(Error::Syntax); // Step 3, 4
+            return Err(Error::Syntax); // Step 4
```
Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: nit: Move the step annotation above the if-conditional.

**Comment by Ms2ger** [comment]
Yeah, we shouldn't try to convert to a string

**Comment by timvandermeij** [comment]
The alternative approach I can think of is to manually loop over the ByteString. Start at the beginning and remove any characters in the whitespace vector until we read the first non-whitespace character, and performing the same process for the end of the string. Or do you know a better approach where we might use existing trim functionality?

**Comment by TimNN** [comment]
I've been think about how to implement this a bit - the problem I see is that `Vec` does currently not allow removing a contiguous sequence of elements, so removing white space from the front would copy the complete data once for every removal.

Before getting your replies in the other thread I've been playing around with this and came up with a utility method on `ByteString` (not sure if this is the best place for the functionality):

```rust
    /// Normalize `self`, as defined by
    /// [the Fetch Spec](https://fetch.spec.whatwg.org/#concept-header-value-normalize).
    pub fn normalize_header_value(&mut self) {
        const HTTP_WS_BYTES: &'static [u8] = b"\x09\x0A\x0D\x20";

        let ByteString(ref mut vec) = *self;
        let back_skip = vec.iter().rev().take_while(|&b| HTTP_WS_BYTES.contains(b)).count();

        if back_skip != 0 {
            let new_len = vec.len() - back_skip;
            vec.truncate(new_len);
        }

        let front_skip = vec.iter().take_while(|&b| HTTP_WS_BYTES.contains(b)).count();

        if front_skip != 0 {
            *vec = vec.split_off(front_skip);
        }
    }
```

I haven't tested it yet but I think this would work and would only allocate if bytes needed to be removed from the front.

**Comment by timvandermeij** [comment]
Your solution looks really good! I like the idea of iterating over the vector like that. It certainly prevents the potential panics in my attempt. Feel free to open a PR for your approach; mine can be closed when you do.

**Comment by timvandermeij** [comment]
Superseded by #9598.

**Pull request closed without merging**